### PR TITLE
feat: add workflow stage resume

### PR DESCRIPTION
**Pull request by andreatnvidia**
## 📋 Summary
Adds stage-level resume support for chained workflows so compatible completed stages can be reused, matching partial stages can continue through the existing single-stage resume path, and downstream stages rerun when upstream outputs change.

## 🔗 Related Issue

N/A

## 🔄 Changes
- Add `CompositeWorkflow.run(resume=...)` with completed-stage reuse and partial-stage delegation.
- Invalidate downstream stages when an upstream stage reruns, changes, or has missing selected/callback output.
- Harden workflow metadata writes and let `ResumeMode.IF_POSSIBLE` fall back to fresh runs when prior metadata is unusable.
- Add public workflow resume tests for skip, rerun, partial/failed-stage resume, callback-output, output-processor, completed-empty, corrupt metadata, and strict resume behavior.
- Document workflow resume behavior in MkDocs and Fern docs.
- Update the workflow chaining plan with the completed stage-level resume slice.

## 🧪 Testing
- [ ] `make test` passes
- [x] Unit tests added/updated
- [ ] E2E tests added/updated (if applicable)

Ran:
- `.venv/bin/ruff format .`
- `.venv/bin/ruff check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/ruff format --check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/pytest packages/data-designer/tests/interface/test_composite_workflow.py -q` - 55 passed, 2 warnings
- `.venv/bin/pytest /home/ubuntu/Code/reviews/DataDesigner-747/smoke_test.py -q -s` - 2 passed against NVIDIA Build (`nvidia/nemotron-3-nano-30b-a3b`) and NVIDIA Inference (`openai/openai/gpt-5.4-nano`) using `/home/ubuntu/Code/.env`

Note: full `.venv/bin/ruff check --fix .` currently hits an unrelated existing generated-notebook lint in `docs/colab_notebooks/7-nemotron-personas.ipynb` (`F404`).

## ✅ Checklist
- [x] Follows commit message conventions
- [ ] Commits are signed off (DCO)
- [x] Architecture docs updated (if applicable)

### CODE REVIEW

File: packages/data-designer/src/data_designer/interface/composite_workflow.py [hunk 1]
```diff
@@ -288,7 +295,43 @@ def run(self) -> CompositeWorkflowResults:
...
+                stage_results[stage.name] = output_result
+                stage_output_paths[stage.name] = output_seed_path
+                previous_seed_path = output_seed_path
+                previous_output_records = output_records
+                previous_stage_name = stage.name
+                previous_stage_fingerprint = stage_fingerprint
+                if stage_metadata["status"] == "completed_empty":
+                    skipped_upstream_stage = stage.name
+                _write_workflow_metadata(workflow_path, metadata)
+                continue
```
[the remark this answers is not in the document]
Comment (andreatnvidia) [reply packages/data-designer/src/data_designer/interface/composite_workflow.py]: Fixed in b767633d: the skipped `completed_empty` path now keeps `previous_output_records` unset instead of carrying `0`, so downstream count state no longer depends on skip-flag ordering.

File: packages/data-designer/tests/interface/test_composite_workflow.py [hunk 2]
```diff
@@ -431,6 +431,188 @@ def test_composite_workflow_clones_stage_builders_on_add(
...
+    metadata_path.write_text(json.dumps(metadata), encoding="utf-8")
+    create_mock.reset_mock()
+
+    resumed = data_designer.compose_workflow(name="resume-partial")
+    resumed.add_stage("base", _category_builder(stub_model_configs), num_records=2)
+    resumed.add_stage("copy", _copy_builder(stub_model_configs))
+    resumed.run(resume=ResumeMode.IF_POSSIBLE)
+
+    assert [call.kwargs["dataset_name"] for call in create_mock.call_args_list] == ["stage-0-base", "stage-1-copy"]
+    assert [call.kwargs["resume"] for call in create_mock.call_args_list] == [ResumeMode.ALWAYS, ResumeMode.NEVER]
```
[the remark this answers is not in the document]
Comment (andreatnvidia) [reply packages/data-designer/tests/interface/test_composite_workflow.py]: Fixed in b767633d: the partial-stage helper now removes completion-only fields from staged metadata before setting `running`/`failed`, so the test matches the interrupted-run metadata shape.

**Comment by johnnygreco** [comment]
Nice work on this one, @andreatgretel — the stage-level resume flow is thoughtful and the test coverage is substantial.

### Summary

This PR adds workflow-level resume support, target-stage materialization, review-gate output overrides, and `export_stage()` for composite workflows, with matching Fern and plan updates. The implementation matches the PR description overall, but there is one metadata hardening edge case worth fixing before merge.

### Findings

#### Warnings — Worth addressing
> Design issues, missing error handling, test gaps, or violations of project standards that could cause problems later.

**`packages/data-designer/src/data_designer/interface/composite_workflow.py:568` — Validate decoded workflow metadata before using it**
- **What**: `_read_prior_workflow_metadata()` assumes `json.loads()` returns a dict. If `workflow-metadata.json` contains syntactically valid but unusable JSON such as `[]`, `null`, or a string, the next `metadata.get(...)` access raises a raw `AttributeError` in both `ResumeMode.IF_POSSIBLE` and `ResumeMode.ALWAYS`. I reproduced this by writing `[]` to `workflow-metadata.json` and calling `workflow.run(resume=ResumeMode.IF_POSSIBLE)`.
- **Why**: The PR is explicitly hardening workflow metadata handling, and `IF_POSSIBLE` is meant to fall back to a fresh run when prior metadata is unusable. A raw `AttributeError` breaks that public resume contract and bypasses the project’s boundary-error normalization.
- **Suggestion**: After decoding, check `isinstance(metadata, dict)`. For `IF_POSSIBLE`, warn and return `None`; for `ALWAYS`, raise `DataDesignerWorkflowError`. A small parametrized test with `[]`, `null`, and `"oops"` would lock this down.

```python
if not isinstance(metadata, dict):
    if resume != ResumeMode.ALWAYS:
        logger.warning("Workflow metadata for %r has invalid shape; starting fresh.", workflow_name)
        return None
    raise DataDesignerWorkflowError(
        f"Cannot resume workflow {workflow_name!r}: workflow metadata has invalid shape."
    )
```

### What Looks Good

- The stage invalidation model is nicely scoped: compatible completed stages are reused, partial stages delegate to single-stage resume, and descendants rerun once an upstream boundary changes.
- The tests cover the important public workflows: completed-stage skip, partial/failed resume delegation, output processors, callback output loss, empty stages, corrupt JSON, `targets`, `rerun_from`, and output overrides.
- Storing workflow-internal paths relatively and writing workflow metadata atomically are both good durability improvements for resumable artifacts.

### Verdict

Needs changes — please add the decoded-metadata shape guard before merge.

---
*This review was generated by an AI assistant.*

**Comment by andreatnvidia** [comment]
Thanks @johnnygreco, good catch. Fixed in `8334456c`.

I added an `isinstance(metadata, dict)` guard immediately after decoding `workflow-metadata.json`.

Behavior now matches the resume contract:
- `ResumeMode.IF_POSSIBLE`: logs a warning and starts fresh for invalid metadata shapes like `[]`, `null`, or `"oops"`
- `ResumeMode.ALWAYS`: raises `DataDesignerWorkflowError`

Added parametrized tests for `[]`, `null`, and `"oops"` in both modes.

Validation:
- `.venv/bin/ruff check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/ruff format --check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/pytest packages/data-designer/tests/interface/test_composite_workflow.py -q` - `67 passed`

**Review by johnnygreco — DISMISSED** [review]
Thanks @andreatgretel, this addresses the metadata-shape edge case cleanly. The added tests cover both resume modes for the invalid JSON shapes we discussed, and my local verification passed. Looks good to me.

**Comment by andreatnvidia** [comment]
@johnnygreco thanks again for the review. I found and fixed one small follow-up before merging: `rerun_from` now requires `resume=ResumeMode.IF_POSSIBLE` or `ResumeMode.ALWAYS`, so it cannot silently behave like a full fresh run with the default `resume=NEVER`.

I also added early validation for `stage_output_overrides` paths so missing or invalid parquet fails before any stages run, and clarified the docs sentence for the `rerun_from` flow.

Validation:
- `.venv/bin/ruff format --check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/ruff check packages/data-designer/src/data_designer/interface/composite_workflow.py packages/data-designer/tests/interface/test_composite_workflow.py`
- `.venv/bin/pytest packages/data-designer/tests/interface/test_composite_workflow.py -q` - `69 passed`

Could you re-approve when you get a chance?

**Review by johnnygreco — APPROVED** [review]
Thanks @andreatgretel, I checked the follow-up changes. The `rerun_from` guard, early `stage_output_overrides` path validation, added tests, and docs clarification all look good to me. Re-approving.
